Confidential-Storage-Arch.md

```diff
@@ -0,0 +1,124 @@
+- Trusted data will be always accessed through trusted filesystems. If users want to directly access data from raw block
+  devices or untrusted filesystems, they must take the responsibility to verify/decrypt those data.
+- Guest memory is a sort of precious resource. It should not be used to store large scale of ephemeral storage content.
+
+## Candidate Technologies and Solutions
+
+With goals and assumptions defined above, the task becomes to define a software stack to convert untrusted storage data
+originated outside of the trusted domain into trusted filesystems within the trusted domain.
+
+### Sources of Untrusted Storage Data
+
+Let's first enumerate the possible sources of untrusted storage data as below:
+- initrd (measured)
+- virtio-blk devices
+- virtio-scsi devices
+- virtio-pmem devices  
+- nvme devices
+- virtio-fs devices
+- virtio-9p devices
+- network attached filesystems, such as NFS, CephFS etc. There's a special case for network attached filesystems, it may
+  be trusted filesystem if it's connected to a trusted server through secure connections.
+- software services providing storage data, for example an image cache service integrated into the image service.  
+
+The storage subsystem may be composed up with multiple layers, so we may enforce confidentiality at different positions.
+Here confidentiality enforcement means decryption and/or integrity verification.
+
+### Enforce Confidentiality at The Data Source Layer
+
+The confidentiality may be enforced at the data source layer, such as at the service agents or the NFS clients.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Data Source Layer](images/storage-enforce-confidentiality-at-source.png)
+
+### Enforce Confidentiality at The Block Device Layer
+
+The confidentiality may also be enforced at the block device layer, such as by using the dm-crypt/dm-verity.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Block Device Layer](images/storage-enforce-confidentiality-at-blockdev.png)
+
+### Enforce Confidentiality at The FileSystem Layer
+
+The confidentiality may also be enforced at the filesystem layer, such as by using the to-be-invented filesystems.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Filesystem Layer](images/storage-enforce-confidentiality-at-fs.png)
+
+### Composition of Confidential Filesystem
+
+Once those confidential filesystems are ready, they could be composed up as normal filesystems in anyway, such as by
+mount, bind mount, overlayfs, chroot or pivot etc.
+
+## Ownership and Life Management of Storage Devices
+
+With traditional Kata Containers, it aims to be compatible with runC and all filesystems are prepared on host, 
+the Kata runtime just pass those storage resources into the guest. With confidential containers, that model doesn't
+work anymore, it needs to prepare filesystems for containers within the vm.
+
```

Comment:
We may want to add a link to the [directly assigned volume work](https://github.com/egernst/kata-containers/blob/da-proposal/docs/design/direct-assign-volume.md) proposal here.